# 死程序不说谎
<!-- 2020.03.30 -->

您是否注意到，有时别人可以在您自己意识到问题之前就发现您的问题？与其他人的代码相同。如果我们的某个程序开始出现问题，首先捕获它的是库或框架例程。也许我们传入了一个 `nil` 值或一个空列表。可能是该哈希中缺少键，或者我们认为包含哈希的值实际上包含一个列表。可能是我们没有发现网络错误或文件系统错误，并且我们的数据为空或损坏。数百万条指令之前的逻辑错误意味着 case 语句的选择器不再是预期的1、2或3。我们将意外地遇到默认大小写。这也是每个 case/switch 语句都需要使用默认子句的原因之一：我们想知道“不可能”发生的时间。

很容易陷入“不可能发生”的想法。我们大多数人编写的代码都无法检查文件是否成功关闭，或者没有按照我们的预期编写跟踪语句。在所有条件都相同的情况下，我们可能不需要这样做—在任何正常情况下，相关代码都不会失败。但是我们是在防御性地编码。我们确保数据就是我们认为的那样，生产中的代码就是我们认为的代码。我们正在检查是否已正确加载依赖项的版本。

所有错误都会为您提供信息。您可以使自己确信该错误不会发生，然后选择忽略它。相反，实用程序员告诉自己，如果有错误，则发生了非常非常糟糕的事情。不要忘记阅读该死的错误消息（请参阅 [_陌生土地上的编码员_](../Chapter3/调试.md) ）

捕获和释放是为了鱼
一些开发人员认为捕获或挽救所有异常，并在编写某种消息后重新引发它们是一种很好的样式。 他们的代码充满了这样的内容（其中一个简单的 raise 语句重新引发了当前异常）：

```elixir
 try do
    add_score_to_board(score);
  rescue InvalidScore
    Logger.error("Can't add invalid score. Exiting");
    raise
  rescue BoardServerDown
    Logger.error("Can't add score: board is down. Exiting");
    raise
  rescue StaleTransaction
    Logger.error("Can't add score: stale transaction. Exiting");
    raise
 end
```

这是实用程序员的写法：

```elixir
  add_score_to_board(score)
```

我们选择它有两个原因。 首先，错误处理不会使应用程序代码黯然失色。 其次，也许更重要的是，代码之间的耦合较少。 在详细示例中，我们必须列出 add_score_to_board 方法可能引发的每个异常。 如果该方法的编写者添加了另一个异常，则我们的代码可能会过时。 在更实用的第二个版本中，新的异常会自动传播。

---
## 提示 38 早期崩溃
---

## 崩溃，不要垃圾

让你尽早发现问题的好处之一就是你可以更早地让程序崩溃。而让程序崩溃通常是您可以做的最好的事情。另一种选择是继续操作，将损坏的数据写入某个重要数据库，或命令洗衣机进入其第二十个连续旋转周期。

Erlang 和 Elixir 语言采用了这种哲学。 Erlang 的发明者，以及《Erlang编程：并发世界的软件》的作者 Joe Armstrong 被经常引用的一句话：“防御性编程是浪费时间。让它崩溃！” 在这些环境中，程序被设计为失败，但是该失败由 supervisor 管理。supervisor 负责运行代码，并且知道在代码失败的情况下该怎么办，这可能包括清理代码，重新启动代码等。supervisor 本身发生故障时会发生什么？它自己的 supervisor 来管理该事件，从而导致由 supervisor 树组成的设计。该技术非常有效，有助于解决这些语言在高可用性，容错系统中的使用。

在其他环境中，仅退出正在运行的程序可能是不合适的。您可能已声明可能无法释放的资源，或者您可能需要编写日志消息，整理未完成的事务或与其他进程进行交互。

但是，基本原理保持不变-当您的代码发现原本不可能发生的事情刚刚发生时，您的程序将不再可行。从现在开始，它所做的任何事情都值得怀疑，因此请尽快将其终止。

失效的程序通常比损坏的程序少得多的损害。

## 相关内容包括
- 话题 25 [_断言式编程_](./断言式编程.md)
- 话题 26 [_如何平衡资源_](./如何平衡资源.md)
- 话题 23 [_按合同设计_](./按合同设计.md)
- 话题 20 [_调试_](../Chapter3/调试.md)
- 话题 43 [_在某处保持安全_](../Chapter7/在某处保持安全.md)
